src/Workspaces/CoreTest/SerializationTests.cs

```diff
@@ -86,5 +89,63 @@ private static void TestSymbolSerialization(Document document, string symbolName
 
             Assert.True(id.Equals(did));
         }
+
+        private static void TextEncodingRoundrip(Encoding encoding)
+        {
+            using var stream = new MemoryStream();
+
+            using (var writer = new ObjectWriter(stream, leaveOpen: true))
+            {
+                SerializerService.WriteTo(encoding, writer, CancellationToken.None);
+            }
+
+            stream.Position = 0;
+
+            using var reader = ObjectReader.TryGetReader(stream);
+            Assert.NotNull(reader);
+            var actualEncoding = (Encoding)SerializerService.ReadEncodingFrom(reader, CancellationToken.None).Clone();
+            var expectedEncoding = (Encoding)encoding.Clone();
+
+            // set the fallbacks to the same instance so that equality comparison does not take them into account:
+            actualEncoding.EncoderFallback = EncoderFallback.ExceptionFallback;
+            actualEncoding.DecoderFallback = DecoderFallback.ExceptionFallback;
+            expectedEncoding.EncoderFallback = EncoderFallback.ExceptionFallback;
+            expectedEncoding.DecoderFallback = DecoderFallback.ExceptionFallback;
+
+            Assert.Equal(expectedEncoding.GetPreamble(), actualEncoding.GetPreamble());
+            Assert.Equal(expectedEncoding.CodePage, actualEncoding.CodePage);
+            Assert.Equal(expectedEncoding.WebName, actualEncoding.WebName);
+            Assert.Equal(expectedEncoding, actualEncoding);
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void EncodingSerialization_UTF8(bool byteOrderMark)
+        {
+            TextEncodingRoundrip(new UTF8Encoding(byteOrderMark));
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void EncodingSerialization_UTF32(bool bigEndian, bool byteOrderMark)
+        {
+            TextEncodingRoundrip(new UTF32Encoding(bigEndian, byteOrderMark));
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void EncodingSerialization_Unicode(bool bigEndian, bool byteOrderMark)
+        {
+            TextEncodingRoundrip(new UnicodeEncoding(bigEndian, byteOrderMark));
+        }
+
+        [Fact]
+        public void EncodingSerialization_AllAvailable()
+        {
+            foreach (var info in Encoding.GetEncodings())
```

Comment:
💡 It would be nice to have this as a `Theory` that uses `MemberData` to provide each encoding by name.